SearchStateExploded exception

This template is a guideline, not a strict requirement.
- **Version**: 2.1.1
- **Integration**: anything
- **Configuration**:
```
version = 2.1.1
```

## Steps

Given code like this:
```scala
object a {
@tailrec
  final def rhsOptimalToken(start: FormatToken): Token = start.right match {
      case T.Comma() | T.LeftParen() | T.RightParen() | T.RightBracket() |
          T.Semicolon() | T.RightArrow() | T.Equals()
          if next(start) != start &&
            !startsNewBlock(start.right) &&
            newlinesBetween(start.between) == 0 =>
        rhsOptimalToken(next(start))
      case _ => start.left
    }
}
```

When I run scalafmt like this:

```bash
scalafmt --stdin
```

## Problem

Scalafmt formats code like this:
```scala
org.scalafmt.cli.FailedToFormat: stdin.scala
Caused by: org.scalafmt.Error$SearchStateExploded: Search state exploded around line 6
        at org.scalafmt.internal.BestFirstSearch.shortestPath(BestFirstSearch.scala:219)
        at org.scalafmt.internal.BestFirstSearch.getBestPath(BestFirstSearch.scala:277)
        at org.scalafmt.Scalafmt$.format(Scalafmt.scala:63)
        at org.scalafmt.cli.ScalafmtCoreRunner$.unsafeHandleFile(ScalafmtCoreRunner.scala:85)
        at org.scalafmt.cli.ScalafmtCoreRunner$.handleFile(ScalafmtCoreRunner.scala:64)
        at org.scalafmt.cli.ScalafmtCoreRunner$.$anonfun$run$3(ScalafmtCoreRunner.scala:42)
        at org.scalafmt.cli.ScalafmtCoreRunner$.$anonfun$run$3$adapted(ScalafmtCoreRunner.scala:41)
        at scala.collection.Iterator.foreach(Iterator.scala:941)
        at scala.collection.Iterator.foreach$(Iterator.scala:941)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
        at scala.collection.parallel.ParIterableLike$Foreach.leaf(ParIterableLike.scala:974)
        at scala.collection.parallel.Task.$anonfun$tryLeaf$1(Tasks.scala:53)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at scala.util.control.Breaks$$anon$1.catchBreak(Breaks.scala:67)
        at scala.collection.parallel.Task.tryLeaf(Tasks.scala:56)
        at scala.collection.parallel.Task.tryLeaf$(Tasks.scala:50)
        at scala.collection.parallel.ParIterableLike$Foreach.tryLeaf(ParIterableLike.scala:971)
        at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.compute(Tasks.scala:153)
        at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.compute$(Tasks.scala:149)
        at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:440)
        at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:189)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
```

## Expectation

Scalafmt doesn't fail

## Notes

Braces around function body solves the issue:

```scala
object a {
 @tailrec
  final def rhsOptimalToken(start: FormatToken): Token = {
    start.right match {
      case T.Comma() | T.LeftParen() | T.RightParen() | T.RightBracket() | T.Semicolon() |
           T.RightArrow() | T.Equals()
          if next(start) != start &&
            !startsNewBlock(start.right) &&
            newlinesBetween(start.between) == 0 =>
        rhsOptimalToken(next(start))
      case _ => start.left
    }
  }
}
```

And this code formats fine:

```scala
object a {
@tailrec
  final def rhsOptimalToken(start: FormatToken): Token = start.right match {
      case Comma() | LeftParen() | RightParen() | RightBracket() |
          Semicolon() | RightArrow() | Equals()
          if next(start) != start &&
            !startsNewBlock(start.right) &&
            newlinesBetween(start.between) == 0 =>
        rhsOptimalToken(next(start))
      case _ => start.left
    }
}
```

--- a/scalafmt-tests/src/test/resources/test/i1527.stat
+++ b/scalafmt-tests/src/test/resources/test/i1527.stat
@@ -1,0 +1,119 @@
+maxColumn=80
+<<< should not throw SearchStateExploded
+object a {
+@tailrec
+  final def rhsOptimalToken(start: FormatToken): Token = start.right match {
+      case T.Comma() | T.LeftParen() | T.RightParen() | T.RightBracket() |
+          T.Semicolon() | T.RightArrow() | T.Equals()
+          if next(start) != start &&
+            !startsNewBlock(start.right) &&
+            newlinesBetween(start.between) == 0 =>
+        rhsOptimalToken(next(start))
+      case _ => start.left
+    }
+}
+>>>
+object a {
+  @tailrec
+  final def rhsOptimalToken(start: FormatToken): Token =
+    start.right match {
+      case T.Comma() | T.LeftParen() | T.RightParen() | T.RightBracket() |
+          T.Semicolon() | T.RightArrow() | T.Equals()
+          if next(start) != start &&
+            !startsNewBlock(start.right) &&
+            newlinesBetween(start.between) == 0 =>
+        rhsOptimalToken(next(start))
+      case _ => start.left
+    }
+}
+<<< scala-native
+def genCQuoteOp(app: Apply): Val = {
+  app match {
+    // Sometimes I really miss quasiquotes.
+    //
+    // case q"""
+    //   scala.scalanative.unsafe.`package`.CQuote(
+    //     new StringContext(scala.this.Predef.wrapRefArray(
+    //       Array[String]{${str: String}}.$asInstanceOf[Array[Object]]()
+    //     ))
+    //   ).c()
+    // """ =>
+    case Apply(
+        Select(
+        Apply(
+        _,
+        List(
+        Apply(_,
+              List(
+              Apply(_,
+                    List(
+                    Apply(TypeApply(
+                          Select(ArrayValue(
+                                 _, List(Literal(Constant(str: String)))),
+                                 _),
+                          _),
+                          _))))))),
+        _),
+        _) =>
+      val chars = Val.Chars(str.replace("\\n", "\n").replace("\\r", "\r"))
+      val const = Val.Const(chars)
+      buf.box(nir.Rt.BoxedPtr, const, unwind)
+
+    case _ =>
+      unsupported(app)
+  }
+}
+>>>
+def genCQuoteOp(app: Apply): Val = {
+  app match {
+    // Sometimes I really miss quasiquotes.
+    //
+    // case q"""
+    //   scala.scalanative.unsafe.`package`.CQuote(
+    //     new StringContext(scala.this.Predef.wrapRefArray(
+    //       Array[String]{${str: String}}.$asInstanceOf[Array[Object]]()
+    //     ))
+    //   ).c()
+    // """ =>
+    case Apply(
+          Select(
+            Apply(
+              _,
+              List(
+                Apply(
+                  _,
+                  List(
+                    Apply(
+                      _,
+                      List(
+                        Apply(
+                          TypeApply(
+                            Select(
+                              ArrayValue(
+                                _,
+                                List(Literal(Constant(str: String)))
+                              ),
+                              _
+                            ),
+                            _
+                          ),
+                          _
+                        )
+                      )
+                    )
+                  )
+                )
+              )
+            ),
+            _
+          ),
+          _
+        ) =>
+      val chars = Val.Chars(str.replace("\\n", "\n").replace("\\r", "\r"))
+      val const = Val.Const(chars)
+      buf.box(nir.Rt.BoxedPtr, const, unwind)
+
+    case _ =>
+      unsupported(app)
+  }
+}

